src/ccl_background.c

```diff
@@ -345,157 +345,168 @@ TASK: if not already there, make a table of comoving distances and of E(a)
+    gsl_spline_free(E); //Note: you are allowed to call gsl_free() on NULL
     gsl_spline_free(chi);
-    *status=CCL_ERROR_MEMORY;
-    ccl_cosmology_set_status_message(cosmo, "ccl_background.c: ccl_cosmology_compute_distances(): ran out of memory\n");
     return;
   }
 
-  a[0]=a0; a[na-1]=af;
+  // Set up the boundaries for the a(chi) spline
+  double dchi=5.;
+  double chi0=chi_a[na-1], chif=chi_a[0];
+  double a0=a[na-1], af=a[0];
+
+  //TODO: The interval in chi (5. Mpc) should be made a macro
+  free(a); //Free these, in preparation for making a(chi) splines
+  free(E_a);
+  free(chi_a);
+  //Note: you are allowed to call free() on NULL
+
+  ////////////////////////////////////////////////////////////////////////////////////////////////////
+  //Below here na (length of some arrays) changes, so this function has to be split at this point.
+  ////////////////////////////////////////////////////////////////////////////////////////////////////
+  na = (int)((chif-chi0)/dchi);
+  dchi  = (chif-chi0)/na; // <=5, since na is an integer
+  //Allocate new arrays for a and chi(a)
+  chi_a = ccl_linear_spacing(chi0, chif, na);
+  a     = malloc(sizeof(double)*na);
+  //Allocate space for GSL root finders
   const gsl_root_fdfsolver_type *T=gsl_root_fdfsolver_newton;
   gsl_root_fdfsolver *s=gsl_root_fdfsolver_alloc(T);
-  for(int i=1;i<na-1;i++) {
-    // we are using the previous value as a guess here to help the root finder
-    // as long as we use small steps in a this should be fine
-    a_of_chi(y[i],cosmo, status, &a0,s);
-    a[i]=a0;
+  gsl_spline * achi=gsl_spline_alloc(A_SPLINE_TYPE, na);
+
+  //Check for too little memory
+  if (!*status){
+    if (a==NULL || chi_a==NULL){
+      *status=CCL_ERROR_MEMORY; 
+      ccl_cosmology_set_status_message(cosmo, "ccl_background.c: ccl_cosmology_compute_distances(): ran out of memory\n");
+    }
   }
 
-  gsl_root_fdfsolver_free(s);
-  if(*status) {
-    free(a);
-    free(y);
-    gsl_spline_free(E);
-    gsl_spline_free(chi);
-    *status = CCL_ERROR_ROOT;
-    ccl_cosmology_set_status_message(cosmo, "ccl_background.c: ccl_cosmology_compute_distances(): a(chi) root-finding error \n");
-    return;
+  //Check for messed up chi conditions
+  if (!*status){
+    if((fabs(chi_a[0]-chi0)>1e-5) || (fabs(chi_a[na-1]-chif)>1e-5)) {
+      *status = CCL_ERROR_LINSPACE; 
+      ccl_cosmology_set_status_message(cosmo, "ccl_background.c: ccl_cosmology_compute_distances(): Error creating linear spacing in chi\n");
+    }
   }
 
-  gsl_spline * achi=gsl_spline_alloc(A_SPLINE_TYPE,na);
-  splstatus=gsl_spline_init(achi,y,a,na);
+  // Calculate a(chi)
+  if (!*status){
+    a[0]=a0; a[na-1]=af;
+    for(int i=1;i<na-1;i++) {
+      // we are using the previous value as a guess here to help the root finder
+      // as long as we use small steps in a this should be fine
+      a_of_chi(chi_a[i],cosmo, status, &a0, s);
+      a[i]=a0;
+    }
+    gsl_root_fdfsolver_free(s);
```

Comment:
This doesn't belong here.  Should put this with the other frees below that happen regardless of status.